networkx/algorithms/d_fragmentation.py

```diff
@@ -0,0 +1,375 @@
+...         ("f", "h"),
+...         ("f", "i"),
+...         ("j", "k"),
+...         ("j", "m"),
+...         ("k", "l"),
+...         ("m", "n"),
+...         ("m", "o"),
+...         ("m", "q"),
+...         ("q", "p"),
+...         ("q", "r"),
+...     ]
+... )
+>>>
+>>> d = 2
+>>> # A set of edges of minimum cardinality
+>>> # that must be deleted to obtain a maximum 2-club from F.
+>>> edge_cut = nx.d_fragmented(F, d)
+>>> edge_cut_list = list(edge_cut)
+>>> sz = len(edge_cut_list)
+>>> sz
+4
+>>> # Since a 2-club cluster subgraph is not unique
+>>> # the actual edge_cut may vary but the size of all
+>>> # minimum edge_cuts should be `sz`.
+
+>>> d = 3
+>>> # A set of edges of minimum cardinality
+>>> # that must be deleted to obtain a maximum 3-club from F.
+>>> edge_cut = nx.d_fragmented(F, d)
+>>> edge_cut_list = list(edge_cut)
+>>> sz = len(edge_cut_list)
+>>> sz
+2
+>>> # Since a 3-club cluster subgraph is not unique
+>>> # the actual edge_cut may vary but the size of all
+>>> # minimum edge_cuts should be `sz`.
+>>> d = 4
+>>> # A set of edges of minimum cardinality
+>>> # that must be deleted to obtain a maximum 4-club from F.
+>>> edge_cut = nx.d_fragmented(F, d)
+>>> edge_cut_list = list(edge_cut)
+>>> sz = len(edge_cut_list)
+>>> sz
+2
+>>> # Since a 4-club cluster subgraph is not unique
+>>> # the actual edge_cut may vary but the size of all
+>>> # minimum edge_cuts should be `sz`.
+>>>
+>>> d = 5
+>>> # A set of edges of minimum cardinality
+>>> # that must be deleted to obtain a maximum 5-club from F.
+>>> # Note that since the diameter of both the components is 5,
+>>> # it is already a 5-club cluster graph.
+>>> # Hence no edges need to be removed from F.
+>>>
+>>> edge_cut = nx.d_fragmented(F, d)
+>>> edge_cut_list = list(edge_cut)
+>>> sz = len(edge_cut_list)
+>>> sz
+0
+
+
+References
+----------
+
+..  [1] Schafer, A. (2009). Diploma Thesis. Friedrich Schiller University Jena Faculty of Mathematics and Computer Science. [Exact Algorithms for s-Club Finding and Related Problems](https://pure.mpg.de/rest/items/item_1587743_4/component/file_1587742/content)
+
+..  [2] Oellerman, O., Patel, A. (2020). The s-club deletion problem on various graph families. (Work in Progress). University of Winnipeg, Canada.
+
+"""
+
+
+from typing import Dict, Generator
+
+import networkx as nx
+from networkx.utils import not_implemented_for
+from networkx.algorithms.tree import recognition
+from networkx.algorithms import matching
+from networkx.algorithms.components import connected
+from operator import itemgetter
+
+
+__all__ = ["d_fragmented"]
+
+
+@not_implemented_for("directed", "multigraph")
+def _is_forest(G: nx.Graph) -> bool:
+    """
+    A local helper that determines if a given graph
+    is a forest (i.e. a collection of pair-wise disconnected trees).
+
+    Parameters
+    ----------
+    G: A networkX graph.
+
+    Returns
+    -------
+    True if G is a forest, False otherwise.
+    """
+    return recognition.is_forest(G)
```

Comment:
Why implement a custom function rather than just using `recognition.is_forest` anywhere where you would use this. There is only one place where it is used and it is decorated by `@not_implemented_for("directed", "multigraph")` anyway

Reply:
I agree with you!

Initially, I thought I might use it multiple times so I wanted a protected wrapper with a friendly name. But I ended up using it only once in the end. That has been corrected now.

Also, I have made some revisions.
**The algorithm suggested for d = 1 for arbitrary graphs is not correct.**

Note that the problem of producing a 1-club cluster graph from a graph G is the same as finding a family of pair-wise disjoint cliques such that the number of edges between the cliques is minimum.

The max weight matching approach is valid for weighted forests with d = 1. To resolve it, I have completely removed the support for graphs that are not forests.

In the future though, more algorithms can be added.